### PR TITLE
Unify most of `SessionConfig` settings into `ConfigOptions`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Towards https://github.com/apache/arrow-datafusion/issues/3887

Also re https://github.com/apache/arrow-datafusion/issues/4349

# Rationale for this change
see https://github.com/apache/arrow-datafusion/issues/3887

TLDR is 
1. A single way to do  configuration options in ConfigOptions makes the code cleaner
2. Using ConfigOptions makes them easier to find (as they are documented)
3. Using ConfigOptions makes these settings dynamically configurable (via environment variable or `SET <variable>` syntax

# What changes are included in this PR?

Move all  `SessionConfig` settings other than the schema / catalog names into the `ConfigOptions` structure,

# Are these changes tested?

# Are there any user-facing changes?

Some fields that were `pub` not must be set via a `with_` method or accessed via an accessor. The number of locations in the DataFusion codebase was quite low, which was a nice surprise to me

The mix of HashMap /
rust struct style / builder styles is somewhat of a bummer. I bet some fancy macro magic could unify the two but I don't know the right incantations

### CODE REVIEW

**Comment by alamb** [comment]
I need to deal with the fact that target_partitions is not a constant (it is a function of the number of cpu cores)

**Comment by alamb** [comment]
FYI @xudong963 and @mvanschellebeeck  -- I am loving the sqllogictest framework

This PR has a good first test -- the output of `SHOW` needs to be normalized (specifically, `target_partitions` isn't a constant value). I have some ideas and plan to follow the `PRAGMA` syntax used by duckdb for passing along options
https://github.com/duckdb/duckdb/blob/85843041ec06fa2d1cafaace993076ef778cdbfe/test/sql/types/nested/list/test_nested_list.test#L5-L10C17

I am working on some code, but I don't expect it will be ready until tomorrow.

**Comment by xudong963** [comment]
> I am working on some code, but I don't expect it will be ready until tomorrow.

Looking forward to it!

File: datafusion/core/tests/sqllogictests/test_files/information_schema.slt [hunk 1]
```diff
@@ -15,6 +15,12 @@
 # specific language governing permissions and limitations
 # under the License.
 
+# target_partitions defaults to num_cores, so set
+# to a known value that is unlikely to be
+# the real number of cores on a system
+statement ok
+SET datafusion.execution.target_partitions=7
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/test_files/information_schema.slt:22]: clever 👍

**Comment by alamb** [comment]
FYI @mingmwang

**Comment by mingmwang** [comment]
@alamb 
Feel free to change to use ConfigOptions directly. If there are any changes need to be made in Ballista accordingly, I will take care.

File: datafusion/core/src/config.rs [hunk 2]
```diff
@@ -27,6 +27,28 @@ use std::env;
...
+    "datafusion.optimizer.repartition_aggregations";
+
+/// Configuration option "datafusion.optimizer.repartition_windows"
+pub const OPT_REPARTITION_WINDOWS: &str = "datafusion.optimizer.repartition_windows";
+
+/// Configuration option "datafusion.execuction_collect_statistics"
+pub const OPT_COLLECT_STATISTICS: &str = "datafusion.execuction_collect_statistics";
+
 /// Configuration option "datafusion.optimizer.filter_null_join_keys"
 pub const OPT_FILTER_NULL_JOIN_KEYS: &str = "datafusion.optimizer.filter_null_join_keys";
```
Comment (thinkharderdev) [inline datafusion/core/src/config.rs:53]: Is there any reason we couldn't just make these an enum? Then we could make the `set_x` and `get_x` methods type safe.

**Review by thinkharderdev — APPROVED** [review]
Really nice!

File: datafusion/core/src/config.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/config.rs:53]: 🤔  that is an excellent idea -- I don't think there is any reason we could not do so. I will file a follow on ticket to do so

**Review by andygrove — APPROVED** [review]

File: datafusion/core/src/config.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/config.rs:53]: https://github.com/apache/arrow-datafusion/issues/4517

**Pull request merged**